nighthawk/test/client_test.cc

```diff
@@ -0,0 +1,148 @@
+#include <chrono>
+
+#include "gtest/gtest.h"
+
+#include "test/mocks/event/mocks.h"
+#include "test/mocks/stats/mocks.h"
+
+#include "test/integration/http_integration.h"
+#include "test/integration/integration.h"
+#include "test/integration/utility.h"
+#include "test/server/utility.h"
+#include "test/test_common/utility.h"
+
+#include "common/api/api_impl.h"
+#include "common/common/thread_impl.h"
+#include "common/filesystem/filesystem_impl.h"
+
+#include "nighthawk/test/mocks.h"
+
+#include "nighthawk/source/client/client.h"
+#include "nighthawk/source/client/factories_impl.h"
+#include "nighthawk/source/client/options_impl.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+namespace Client {
+
+class ClientTest : public Envoy::BaseIntegrationTest,
+                   public testing::TestWithParam<Envoy::Network::Address::IpVersion> {
+public:
+  ClientTest() : Envoy::BaseIntegrationTest(GetParam(), realTime(), ClientTest::envoy_config) {}
+
+  static void SetUpTestCase() {
+    Envoy::Filesystem::InstanceImplPosix file_system;
+    envoy_config = file_system.fileReadToEnd(Envoy::TestEnvironment::runfilesPath(
+        "nighthawk/test/test_data/benchmark_http_client_test_envoy.yaml"));
+    envoy_config = Envoy::TestEnvironment::substitute(envoy_config);
+  }
+
+  void SetUp() override {
+    // We fork the integration test fixture into a child process, to avoid conflicting
+    // runtimeloaders as both NH and the integration server want to own that and we can have only
+    // one.
+    pipe(fd_port);
+    pipe(fd_confirm);
+    pid_ = fork();
+    ASSERT(pid_ >= 0);
+
+    if (pid_ == 0) {
+      // child process running the integration test server.
+      ares_library_init(ARES_LIB_INIT_ALL);
+      Envoy::Event::Libevent::Global::initialize();
+      initialize();
+      int port = lookupPort("listener_0");
+      int parent_message;
+      write(fd_port[1], &port, sizeof(port));
+      // The parent process writes to fd_confirm when it has read the port. This call to read blocks
+      // until that happens.
+      read(fd_confirm[0], &parent_message, sizeof(parent_message));
+      ASSERT(parent_message == 1234);
+      // The parent process closes fd_port when the test tears down. The read call blocks until it
+      // does that.
+      ASSERT(read(fd_port[0], &port_, sizeof(port_)) == -1);
+    } else if (pid_ > 0) {
+      ASSERT(read(fd_port[0], &port_, sizeof(port_)) > 0);
+      ASSERT(port_ > 0);
+      int foo = 1234;
+      write(fd_confirm[1], &foo, sizeof(foo));
+    }
+  }
+
+  void TearDown() override {
+    if (pid_ == 0) {
+      test_server_.reset();
+      fake_upstreams_.clear();
+      ares_library_cleanup();
+    }
+    close(fd_confirm[0]);
```

Comment:
Kind of really not a fan of mixing gtest in parent and child, since we now have two interleaved programs. I really want something more like Python's subprocess. Envoy has a pretty good way of discovering its port even when binding admin and listener to 0. The admin port can be discovered via a CLI flag that dumps to a file, the listeners can be pulled from this. See https://github.com/envoyproxy/envoy/blob/d8e0f9c1cd3ad3ed6fc932d3033a6d29e62d3fa8/test/integration/hotrestart_test.sh#L71

Reply:
+1, it's on the tracking issue. I had not foreseen running into the singleton/runtime conflict issue with the integration test server.